contractcourt/htlc_timeout_resolver.go

```diff
@@ -733,3 +740,163 @@ func (h *htlcTimeoutResolver) HtlcPoint() wire.OutPoint {
+// the resolver itself quits. When a spend event is received from the mempool,
+// however, it won't exit but continuing to wait for a spend event from the
+// block subscription.
+//
+// NOTE: there could be a case where we found the preimage in the mempool,
+// which will be added to our preimage beacon and settle the incoming link.
+// Meanwhile, the timeout sweep tx is confirmed, which should have caused the
+// incoming link to fail the payment. This is likely to be an attack and we
+// will play safe here - claim the htlc from upstream using the preimage and
+// race against the downstream for the onchain htlc via timeout path.
+func (h *htlcTimeoutResolver) consumeSpendEvents(resultChan chan *spendResult,
+	blockSpent, mempoolSpent <-chan *chainntnfs.SpendDetail) {
+
+	op := h.HtlcPoint()
+
+	// Create a result chan to hold the results.
+	result := &spendResult{}
+
+	// Wait for a spend event to arrive.
+	for {
+		select {
+		// If a spend event is received from the block, this outgoing
+		// htlc is spent either by the remote via the preimage or by us
+		// via the timeout. We can exit the loop and `claimCleanUp`
+		// will feed the preimage to the beacon if found. This treats
+		// the block as the final judge and the preimage spent won't
+		// appear in the mempool afterwards.
+		//
+		// NOTE: if a reorg happens, the preimage spend can appear in
+		// the mempool again. Though a rare case, we should handle it
+		// in a dedicated reorg system.
+		case spendDetail, ok := <-blockSpent:
+			if !ok {
+				result.err = fmt.Errorf("block spent err: %w",
+					errResolverShuttingDown)
+			} else {
+				log.Debugf("Found confirmed spend of HTLC "+
+					"output %s in tx=%s", op,
+					spendDetail.SpenderTxHash)
+
+				result.spend = spendDetail
+
+				// Once confirmed, persist the state on disk.
+				result.err = h.checkPointSecondLevelTx()
+			}
+
+			// Send the result and exit the loop.
+			resultChan <- result
+
+			return
+
+		// If a spend event is received from the mempool, this can be
+		// either the 2nd stage timeout tx or a preimage spend from the
+		// remote. We will further check whether the spend reveals the
+		// preimage and add it to the preimage beacon to settle the
+		// incoming link.
+		//
+		// NOTE: we won't exit the loop here so we can conitnue to
+		// watch for the block spend to check point the resolution.
+		case spendDetail, ok := <-mempoolSpent:
+			if !ok {
+				result.err = fmt.Errorf("mempool spent err: %w",
+					errResolverShuttingDown)
+
+				// This is an internal error so we exit.
+				resultChan <- result
+
+				return
+			}
+
+			log.Debugf("Found mempool spend of HTLC output %s "+
+				"in tx=%s", op, spendDetail.SpenderTxHash)
+
+			// Check whether the spend reveals the preimage, if not
+			// conitnue the loop.
+			hasPreimage := isPreimageSpend(
+				spendDetail,
+				h.htlcResolution.SignedTimeoutTx != nil,
+			)
+			if !hasPreimage {
+				log.Debugf("HTLC output %s spent doesn't "+
```

Comment:
IIUC, this is expected when we're the ones that sweep already, though the resolver should still exit after everything has resolved, which'll then tear down this goroutine.

Reply:
yeah the debug log was added to better view the 2nd timeout tx being caught by the mempool notifier.